### PR TITLE
Targeting Optimizations

### PR DESCRIPTION
**Pull request by TaranDahl**
1. Technos won't SelectAutoTarget if they are not IsArmed. Didn't check thoroughly, more tests or disassembling are needed.
2. New flags to customize the scanning delay per techno, and differently for player and AI. Default to the vanilla flags `NormalTargetingDelay` and `GuardAreaTargetingDelay`.
As enemy scanning is a costly part of the total overhead, this should be able to improve the performance, especially on the maps with numbers of enemies.
```
[General] or [SomeTechno]
AINormalTargetingDelay=
PlayerNormalTargetingDelay=
AIGuardAreaTargetingDelay=
PlayerGuardAreaTargetingDelay=
```
P.S.: The docs are coming, s∞n.
image

### CODE REVIEW

File: src/Ext/Techno/Hooks.Misc.cpp [hunk 1]
```diff
@@ -98,3 +98,63 @@ DEFINE_HOOK(0x6B7600, SpawnManagerClass_AI_InitDestination, 0x6)
...
+
+DEFINE_HOOK(0x70989C, TechnoClass_TargetAndEstimateDamage_ScanDelayNormal, 0x6)
+{
+	GET(TechnoClass*, pThis, ESI);
+
+	auto const pTypeExt = TechnoTypeExt::ExtMap.Find(pThis->GetTechnoType());
+	auto const pOwner = pThis->Owner;
+	auto const pRulesExt = RulesExt::Global();
+	auto const pRules = RulesClass::Instance();
+	int delay = (pThis->Location.X + pThis->Location.Y + Unsorted::CurrentFrame) % 3;
```
Comment (MortonPL) [inline src/Ext/Techno/Hooks.Misc.cpp:146]: Why does this use the unit's location?

Comment (TaranDahl) [reply src/Ext/Techno/Hooks.Misc.cpp:156]: A simple trick to generate randomness. As the targeting function is called frequently, it may save a little performance overhead compared to the RandomRanged function. I think.

Comment (TaranDahl) [reply src/Ext/Techno/Hooks.Misc.cpp:156]: If not necessary, I can replace it with RandomRanged.

**Comment by TaranDahl** [comment]
> Documentation please image

It's here now.

File: src/Ext/Techno/Hooks.Misc.cpp (hunk 1 above)
Comment (Metadorius) [reply src/Ext/Techno/Hooks.Misc.cpp:156]: IMO better use the random. Just for the clarity, because people are used to the random calls, and like Morton they will ask "what is this" when stumble upon this.

File: CREDITS.md [hunk 2]
```diff
@@ -367,6 +367,11 @@ This page lists all the individual contributions to the project by their author.
   - Skirmish AI "sell all buildings and set all technos to hunt" behavior dehardcode
   - Skirmish AI "gather when MCV deploy" behavior dehardcode
   - Global value of `RepairBaseNodes` and `MCVRedeploys`
+  - Target scanning delay customization.(codes)
+  - Skip target scanning function calling for unarmed technos.(codes)
+  **solar-III (凤九歌)**
+  - Target scanning delay customization.(docs)
+  - Skip target scanning function calling for unarmed technos.(docs)
```
Comment (MortonPL) [inline CREDITS.md:374]:
```suggestion
  - Target scanning delay customization (code)
  - Skip target scanning function calling for unarmed technos (code)
**solar-III (凤九歌)**
  - Target scanning delay customization (documentation)
  - Skip target scanning function calling for unarmed technos (documentation)
```

File: docs/Whats-New.md [hunk 3]
```diff
@@ -529,6 +529,7 @@ Vanilla fixes:
 - Fixed some locomotors (Tunnel, Walk, Mech) getting stuck when moving too fast (by NetsuNegi)
 - Animations with `MakeInfantry` and `UseNormalLight=false` that are drawn in unit palette will now have cell lighting changes applied on them (by Starkku)
 - Fixed Nuke & Dominator Level lighting not applying to AircraftTypes (by Starkku)
+- Skip target scanning function calling for unarmed technos.(by TaranDahl and solar-III)
```
Comment (MortonPL) [inline docs/Whats-New.md:532]:
```suggestion
- Skip target scanning function calling for unarmed technos (by TaranDahl and solar-III)
```

File: src/Ext/Techno/Hooks.Misc.cpp (hunk 1 above)
Comment (TaranDahl) [reply src/Ext/Techno/Hooks.Misc.cpp:156]: Alright I will change it.

File: docs/Fixed-or-Improved-Logics.md [hunk 4]
```diff
@@ -1391,6 +1392,25 @@ In `rulesmd.ini`:
 IsSingleColor=false  ; boolean
 ```
 
+### Target scanning delay optimization
+In Vanilla, the game used 'NormalTargetingDelay' and 'GuardAreaTargetingDelay' to globally control the target searching intervals. Increasing these values would make units stupid, while decreasing them would cause the game lagging.
+Now, you can define them per techno, also allowing different values for AI and players. The default values are the same as those originally defined by the vanilla flags.
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:1397]:
```suggestion
### Target scanning delay optimization

- In vanilla, the game used 'NormalTargetingDelay' and 'GuardAreaTargetingDelay' to globally control the target searching intervals. Increasing these values would make units stupid, while decreasing them would cause game lag.
- Now, you can define them per techno, also allowing different values for AI and players. The default values are the same as those originally defined by the vanilla flags.
```

File: docs/Fixed-or-Improved-Logics.md [hunk 5]
```diff
@@ -1391,6 +1392,25 @@ In `rulesmd.ini`:
...
+PlayerNormalTargetingDelay=          ;integer, game frames
+AIGuardAreaTargetingDelay=           ;integer, game frames
+PlayerGuardAreaTargetingDelay=       ;integer, game frames
+
+[SOMETECHNO]
+AINormalTargetingDelay=              ;integer, game frames
+PlayerNormalTargetingDelay=          ;integer, game frames
+AIGuardAreaTargetingDelay=           ;integer, game frames
+PlayerGuardAreaTargetingDelay=       ;integer, game frames
+```
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:1412]:
```suggestion
In `rulesmd.ini`:
```ini
[General]
AINormalTargetingDelay=              ; integer, game frames
PlayerNormalTargetingDelay=          ; integer, game frames
AIGuardAreaTargetingDelay=           ; integer, game frames
PlayerGuardAreaTargetingDelay=       ; integer, game frames

[SOMETECHNO]                         ; TechnoType
AINormalTargetingDelay=              ; integer, game frames
PlayerNormalTargetingDelay=          ; integer, game frames
AIGuardAreaTargetingDelay=           ; integer, game frames
PlayerGuardAreaTargetingDelay=       ; integer, game frames
```
```

**Review by MortonPL — COMMENTED** [review]
This can be merged after tests.

**Comment by TaranDahl** [comment]
> This can be merged after tests.

Some friends in Chinese community tested this. No fatal error, desync, or misbehavior was found.
Didn't check for edge cases carefully, though.

**Review by MortonPL — APPROVED** [review]

File: docs/Fixed-or-Improved-Logics.md [hunk 6]
```diff
@@ -1215,7 +1216,7 @@ Ammo.DeployUnlockMaximumAmount=-1    ; integer
 ```
 
 ```{warning}
-Auto-deploy feature requires `Convert.Deploy` from [Ares’ Type Conversion](https://ares-developers.github.io/Ares-docs/new/typeconversion.html) to change type. Unit without it will constantly use deploy command on self until ammo is changed.
+Auto-deploy feature requires `Convert.Deploy` from [Ares 1�7 Type Conversion](https://ares-developers.github.io/Ares-docs/new/typeconversion.html) to change type. Unit without it will constantly use deploy command on self until ammo is changed.
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:1219]:
```suggestion
Auto-deploy feature requires `Convert.Deploy` from [Ares' Type Conversion](https://ares-developers.github.io/Ares-docs/new/typeconversion.html) to change type. Unit without it will constantly use deploy command on self until ammo is changed.
```

File: docs/Fixed-or-Improved-Logics.md [hunk 7]
```diff
@@ -1499,6 +1500,24 @@ In `rulesmd.ini`
 FireOnce.ResetSequence=true  ; boolean
 ```
 
+### Force techno targeting in distributed frames to improve performance
+
+- When you create many techno in a same frame (f.ex. starting the game with a campaign map that initially has a large number of technos), they will always scan for target in a synchronous period, causing the game lagging. Increasing targeting delay will not make things better, as their targeting are still synchronized.
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:1505]:
```suggestion
- When you create many technos in a same frame (i.e. starting the game with a campaign map that initially has a large number of technos), they will always scan for targets in a synchronous period, causing game lag. Increasing targeting delay alone will not make things better, as their targeting is still synchronized.
```

File: docs/Fixed-or-Improved-Logics.md [hunk 8]
```diff
@@ -1499,6 +1500,24 @@ In `rulesmd.ini`
...
+- It is now possible to force them targeting seperatly. When a techno spawn, they generate a random number in \[0,15\].
+- If `DistributeTargetingFrame=true` is set, only when the current frame number is congruent with the technos own number under modulo 16, will it do targeting.
+- You can use `DistributeTargetingFrame.AIOnly` to make it only work for AI (Players are not likely to have so many technos.).
+- For technos with low ROF (f.ex. Tanya), this will significantly reduce their efficiency in killing enemies. Turn the same flag off on techno to prevent it.
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:1509]:
```suggestion
- It is now possible to force them to seek targets separately. When a techno spawns, it will generate a random number in \[0,15\]. If `DistributeTargetingFrame=true` is set, only when the current frame number is congruent with the technos own number under modulo 16, will it do targeting.
  - You can use `DistributeTargetingFrame.AIOnly` to make it only work for AI (Players are not likely to have so many technos.)
  - For technos with low ROF (i.e. Tanya), this will significantly reduce their efficiency in killing enemies.
```

File: docs/Fixed-or-Improved-Logics.md [hunk 9]
```diff
@@ -1499,6 +1500,24 @@ In `rulesmd.ini`
...
+
+In `rulesmd.ini`
+```ini
+[General]
+DistributeTargetingFrame=false         ; boolean
+DistributeTargetingFrame.AIOnly=true   ; boolean
+
+[SomeTechno]
+DistributeTargetingFrame=              ; boolean
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:1518]:
```suggestion
[SOMETECHNO]                           ; TechnoType
DistributeTargetingFrame=              ; boolean
```

**Review by MortonPL — COMMENTED** [review]
Just improving the documentation.

**Comment by mevitar** [comment]
Deployed infantry appear to be ignoring these tags. In fact, they also appear to be ignoring the two original tags. They shoot immediately at anything that gets nearby, no matter how high delay values i set.

Didn't notice any other issues.

**Review by CrimRecya — COMMENTED** [review]
From my personal perspective, the implementation of `DistributeTargetingFrame` may not be as efficient as modifying the random range of `TargetingTimer` which is default to (0,2).

**Comment by TaranDahl** [comment]
> From my personal perspective, the implementation of `DistributeTargetingFrame` may not be as efficient as modifying the random range of `TargetingTimer` which is default to (0,2).

Yeah. It does incur some additional overhead. However, its purpose is not to replace that randomness. It is designed to address the frame rate drops caused by the periodic auto-targeting of a large number of units pre-placed on the map.

**Comment by CrimRecya** [comment]
> Yeah. It does incur some additional overhead. However, its purpose is not to replace that randomness. It is designed to address the frame rate drops caused by the periodic auto-targeting of a large number of units pre-placed on the map.

If that's the case, then I think we can try to make `TargetingTimer` start a random value during `TechnoClass::Init`?

**Review by CrimRecya — APPROVED** [review]
